File: 0797_AllPathsFormSourceToTarget/797.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {

    typedef vector<vector<int>> Graph;
    int endPoint;

public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        endPoint = graph.size() - 1;
        Graph res;
        vector<int> path;
        dfs(graph, 0, path, res);
        return res;
    }

    void dfs(Graph& graph, int point, vector<int>& path, Graph& res) {
        path.emplace_back(point);

        if (point == endPoint) {
            res.emplace_back(path);
        } else {
            for (int p: graph[point]) {
                dfs(graph, p, path, res);
            }
        }

        path.pop_back();
    }
};

int main() {
    Solution s = Solution();
    vector<vector<int>> q = { {1,2},{3},{3},{} };
    auto res = s.allPathsSourceTarget(q)
    return 0;
}

File: 0797_AllPathsFormSourceToTarget/797.swift

class Solution {
    
    func allPathsSourceTarget(_ graph: [[Int]]) -> [[Int]] {
        
        var endPoint: Int = graph.count - 1
        
        var res: [[Int]] = []
        
        var path: [Int] = []
        
        func dfs(point: Int) {
            path.append(point)
            defer { path.removeLast() }
            if point == endPoint {
                res.append(path)
                return
            }
            
            for p in graph[point] {
                dfs(point: p)
            }
        }
        
        dfs(point: 0)
        
        return res
    }
}

